zerver/lib/request.py

```diff
@@ -24,6 +27,43 @@
 
 from zerver.lib.exceptions import ErrorCode, InvalidJSONError, JsonableError
 from zerver.lib.types import Validator, ViewFuncT
+from zerver.models import Client, Realm
+
+
+@dataclass
+class ZulipRequestData:
+    """This contains the additional attributes for a request that we
+    add in zerver/lib/middleware.py. Presumably, certain fields are
+    expected to not None in `request_data_map[request]` as long as
+    the request is being accessed after being processed in the middleware.
+
+    """
+
+    client: Optional[Client] = None
+    client_name: Optional[str] = None
+    client_version: Optional[str] = None
+    _log_data: Optional[MutableMapping[str, Any]] = None
+    _rate_limit: Optional[str] = None
+    _requestor_for_logs: Optional[str] = None
+    realm: Optional[Realm] = None
+    _set_language: Optional[str] = None
+    _ratelimits_applied: Optional[List[Any]] = None
+    _query: Optional[str] = None
+    error_format: Optional[str] = None
+    placeholder_open_graph_description: Optional[str] = None
+    saved_response: Optional[HttpResponse] = None
+    _tornado_handler: Any = None
+
+
+request_data_map: MutableMapping[HttpRequest, ZulipRequestData] = weakref.WeakKeyDictionary()
+
+
+def get_or_create_request_data(request: HttpRequest) -> ZulipRequestData:
```

Comment:
I know I suggested “request data”, but thinking more carefully, that will probably get confused with POST data. A more unique name might work better—maybe “request notes”?

Also, shortening `get_or_create` to just `get` should be fine. (One could imagine that it conceptually always exists, and lazy creation is just an internal optimization.)

Reply:
`ZulipRequestNotes` and `get_request_notes` sound correct to me.